firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4537,31 +4536,76 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
     IntegrationTestHelpers.waitFor(semaphore);
   }
 
-  private static FirebaseApp appForDatabaseUrl(String url, String name) {
-    return FirebaseApp.initializeApp(
-        InstrumentationRegistry.getInstrumentation().getTargetContext(),
-        new FirebaseOptions.Builder()
-            .setApplicationId("appid")
-            .setApiKey("apikey")
-            .setDatabaseUrl(url)
-            .build(),
-        name);
+  @Test
+  public void testGetPropagatesEventsToListeners()
+      throws DatabaseException, InterruptedException, ExecutionException {
+    /**
+     * There's a situation where if you have a pending `get()` and then an `onValue()` at the same
+     * location and get resolves before onValue does, onValue gets no events. Note: This is a smoke
+     * test due to the fact that it is difficult to reproduce the scenario when the parent get()
+     * request resolves before a child onValue listener.
+     */
+    // To ensure that we don't read the cached results, we need a separate app.
+    AtomicBoolean successful = new AtomicBoolean();
+    int count = 0;
+    while (count++ < 5 && !successful.get()) {
+
+      FirebaseDatabase readDb = IntegrationTestHelpers.getNewDatabase();
+      FirebaseDatabase writeDb = IntegrationTestHelpers.getNewDatabase();
+      DatabaseReference writeNode = writeDb.getReference().push();
+      DatabaseReference readNode = IntegrationTestHelpers.translateReference(writeNode, readDb);
+      try {
+        Map<String, Object> toWrite =
+            new MapBuilder().put("foo1", "child1").put("foo2", "child2").build();
+        await(writeNode.setValue(toWrite));
+        readDb.goOffline();
+        AtomicBoolean resolved = new AtomicBoolean();
+        Task<DataSnapshot> getTask = readNode.get();
+        getTask.addOnCompleteListener(
+            new OnCompleteListener<DataSnapshot>() {
+              @Override
+              public void onComplete(@NonNull Task<DataSnapshot> task) {
+                DataSnapshot snapshot = task.getResult();
+                assertEquals(toWrite, snapshot.getValue());
+                resolved.set(true);
+              }
+            });
+
+        DatabaseReference childRef = readNode.child("foo1");
+        readDb.goOnline();
```

Comment:
Why not move the `goOnline` call to right after we create the `ReadFuture`? That way we don't need to have an explicit thread sleep.